### ISSUE TITLE
kvserver: lease on-disk and in-memory state diverged

### ISSUE THREAD

**Issue by erikgrinaker** [cockroachdb/cockroach#101896]
Seen on a 23.1 test cluster:

```
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514  on-disk and in-memory state diverged: [Lease.Start.WallTime: 0 != 1681948488725657048 Lease.Replica.NodeID: 0 != 8 Lease.Replica.StoreID: 0 != 8 Lease.Replica.ReplicaID: 0 != 1 Lease.ProposedTS: nil != &hlc.ClockTimestamp{WallTime:1681948488727796023, Logical:0, Synthetic:false} Lease.Epoch: 0 != 2 Lease.Sequence: 0 != 2719 Lease.AcquisitionType: 0 != 2 TruncatedState.Index: 0 != 29 TruncatedState.Term: 0 != 6 Version: nil != &roachpb.Version{Major:22, Minor:2, Patch:0, Internal:0}]
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !goroutine 668 [running]:
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/util/log.getStacks(0x1)
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/util/log/get_stacks.go:25 +0x89
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/util/log.(*loggerT).outputLogEntry(0xc000c3cc00, {{{0xc012ef0090, 0x24}, {0x0, 0x0}, {0x5ad8823, 0x1}, {0x5ad8824, 0x1}}, 0x17577b3d63af3240, ...})
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/util/log/clog.go:262 +0x97
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/util/log.logfDepthInternal({0x6fa51a0, 0xc0222089c0}, 0x2, 0x4, 0x0, 0x0?, {0x5b03315, 0x28}, {0xc025915c38, 0x1, ...})
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:106 +0x645
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/util/log.logfDepth(...)
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/util/log/channels.go:39
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/util/log.Fatalf(...)
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/bazel-out/k8-opt/bin/pkg/util/log/log_channels_generated.go:848
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).assertStateRaftMuLockedReplicaMuRLocked(0xc01e2b4580, {0x6fa51a0, 0xc0222089c0}, {0x6fe4ce0, 0xc000705500})
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica.go:1519 +0x50e
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaStateMachine).ApplySideEffects(0xc01e2b4680, {0x6fa51a0, 0xc0222089c0}, {0x6fe7730?, 0xc02c483408})
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_application_state_machine.go:203 +0x78d
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.mapCheckedCmdIter({0x7f2726ba9a48?, 0xc01e2b4a50}, 0xc0259167d8)
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/cmd.go:210 +0x158
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).applyOneBatch(0xc025916cd0, {0x6fa51a0, 0xc0222089c0}, {0x6fd6a88, 0xc01e2b49f0})
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:295 +0x1f8
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).ApplyCommittedEntries(0xc025916cd0, {0x6fa51a0, 0xc0222089c0})
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:251 +0x9a
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:996 +0x13b9
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady(_, {_, _}, {{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...})
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_raft.go:709 +0x17b
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady(0xc000e9e000, 0x6fa51a0?)
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_raft.go:645 +0x145
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftSchedulerShard).worker(0xc016781950, {0x6fa51a0, 0xc00d1c8600}, {0x6f89ac0, 0xc000e9e000}, 0xc016938400)
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:392 +0x1f2
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).Start.func2({0x6fa51a0?, 0xc00d1c8600?})
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/scheduler.go:299 +0x45
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:470 +0x146
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
F230419 23:54:49.441817 668 kv/kvserver/replica.go:1519 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16514 !	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:461 +0x43b
```

Jira issue: CRDB-27176

**Comment by tbg** [comment]
I downloaded the logs and a debug.zip, will put them on gdrive

Basic timeline: replica `/4` removed at `23:54:49.26`

```
I230419 23:54:49.267726 662 kv/kvserver/store_remove_replica.go:151 ⋮ [T1,n2,s2,r174164/4:‹/Table/738/1/5{39028…-42713…}›,raft] 16488  removing replica r174164/4
```

Snapshot for `/6` (sent by n8) applied at `23:54:49.33`

I230419 23:54:49.338996 4683239 kv/kvserver/replica_raftstorage.go:489 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›] 16499  applied INITIAL snapshot e344ed7c from (n8,s8):1 at applied index 29 (total=28ms data=4.7 KiB ingestion=6@22ms)

Explodes on next raft ready handling at `23:54:49.44` because the state read from disk comes back empty.

```
E230419 23:54:49.441535 668 kv/kvserver/replica.go:1516 ⋮ [T1,n2,s2,r174164/6:‹/Table/738/1/5{39028…-42713…}›,raft] 16509 +[
‹Lease.Start.WallTime: 0 != 1681948488725657048› 
‹Lease.Replica.NodeID: 0 != 8› 
‹Lease.Replica.StoreID: 0 != 8› 
‹Lease.Replica.ReplicaID: 0 != 1› 
‹Lease.ProposedTS: nil != &hlc.ClockTimestamp{WallTime:1681948488727796023, Logical:0, Synthetic:false}› 
‹Lease.Epoch: 0 != 2› 
‹Lease.Sequence: 0 != 2719› 
‹Lease.AcquisitionType: 0 != 2› 
‹TruncatedState.Index: 0 != 29› 
‹TruncatedState.Term: 0 != 6› 
‹Version: nil != &roachpb.Version{Major:22, Minor:2, Patch:0, Internal:0}›]
```

Going to poke at the data dir for n2 to see if the snapshot data is present or not.

**Comment by pavelkalinnikov** [comment]
```
$ ./cockroach debug range-descriptors /mnt/data1/cockroach-lease-state-fatal/ | grep -C 1 "r174164"
1681948489.341418018,0 /Local/Range/Table/738/1/5390282359200842992/RangeDescriptor (0x016b12f702e289fd4ace213f2b3884f00001726473630017577b3d5db33a2209): [/Table/738/1/5390282359200842992, /Table/738/1/5427138990716746192)
        Raw:r174164:/Table/738/1/5{390282359200842992-427138990716746192} [(n8,s8):1, (n3,s3):5VOTER_DEMOTING_LEARNER, (n6,s6):3, (n2,s2):6VOTER_INCOMING, next=7, gen=11936, sticky=9223372036.854775807,2147483647]
```

**Comment by tbg** [comment]
We have the range applied state, it's at raft log index 31, and we got the snapshot at 29, indicating that we applied 30 and 31 before crashing. We can see that index 30 was a "normal" write and index 31 was a conf change, which is likely what triggered the assertion check (we compare disk and mem state only after "interesting" commands). So this all checks out.

We have the `rftr` (RaftLocalReplicaID) key and it stores a replicaID of 6 (it's printed here as a `raftpb.Entry`, but it is the right thing, I checked[^1]).

[^1]: ```
func TestFoo(t *testing.T) {
	id := RaftReplicaID{ReplicaID: 6}
	slID, err := id.Marshal()
	require.NoError(t, err)
	var ent raftpb.Entry
	require.NoError(t, ent.Unmarshal(slID))
	t.Log(ent)
}
````

- `0,0 /Local/RangeID/174164/r/RangeAppliedState` (0x0169f802a854727261736b00): raft_applied_index:31 lease_applied_index:20 range_stats:<last_update_nanos:1681948488946755650 sys_bytes:4920 sys_count:15 > raft_closed_timestamp:<wall_time:1681948486388512825 > raft_applied_index_term:6
- `0,0 /Local/RangeID/174164/u/RangeTombstone` (0x0169f802a854757266746200): Term:0 Index:0 Type:6 : EMPTY
- `0,0 /Local/RangeID/174164/u/RaftHardState` (0x0169f802a854757266746800): Term:0 Index:31 Type:6 : EMPTY
- `0,0 /Local/RangeID/174164/u/RaftLog/logIndex:30` (0x0169f802a854757266746c000000000000001e00): Term:6 Index:30 Type:EntryNormal  (ID 6edaf0eeec3ad1ee)
- `0,0 /Local/RangeID/174164/u/RaftLog/logIndex:31` (0x0169f802a854757266746c000000000000001f00): Term:6 Index:31 Type:EntryConfChangeV2  (ID 10be28fb21d9cb00)
- `0,0 /Local/RangeID/174164/u"rftr"` (0x0169f802a854757266747200): Term:0 Index:0 Type:6 : EMPTY
- `1681948489.341418018,0 /Local/Range/Table/738/1/5390282359200842992/RangeDescriptor` (0x016b12f702e289fd4ace213f2b3884f00001726473630017577b3d5db33a2209): [/Table/738/1/5390282359200842992, /Table/738/1/5427138990716746192)
	Raw:r174164:/Table/738/1/5{390282359200842992-427138990716746192} [(n8,s8):1, (n3,s3):5VOTER_DEMOTING_LEARNER, (n6,s6):3, (n2,s2):6VOTER_INCOMING, next=7, gen=11936, sticky=9223372036.854775807,2147483647]
- `0,0 /Local/Range/Table/738/1/5390282359200842992/Transaction/"cf25d022-db83-4254-84f1-57651b5550aa"` (0x016b12f702e289fd4ace213f2b3884f0000174786e2dcf25d022db83425484f157651b5550aa00): meta={id=cf25d022 key=/Local/Range/Table/738/1/5390282359200842992/RangeDescriptor pri=0.00009718 epo=0 ts=1681948489.341418018,0 min=1681948489.341418018,0 seq=15} lock=true stat=COMMITTED rts=0,0 wto=false gul=0,0 int=13

**Comment by tbg** [comment]
The pebble logs end too early:

```
I230419 23:54:48.672823 4681502 3@pebble/event.go:694 ⋮ [n2,s2,pebble] 101583  [JOB 32665] flushing 6 ingested tables
I230419 23:54:48.676487 4681502 3@pebble/event.go:698 ⋮ [n2,s2,pebble] 101584  [JOB 32665] flushed 6 ingested flushables L0:2104850 (1.4 K) + L0:2104855 (1.2 K) + L0:2104851 (1.8 K) + L0:2104852 (1.2 K) + L0:2104853 (1.2 K) + L6:2104854 (1.2 K) in 0.0s (0.0s total), output rate 4.6 M/s
(END)
```

Ingesting six SSTs sounds like a snapshot application, but our snapshot only got initiated >500ms later, at 23:54:49. So this logging is for another snapshot on n2, but our range was only split off at `23:54:48.94`, though it did get a snapshot at `23:54:49.084506` at log index 13. That snapshot is also not logged in pebble.

It's surprising to me that the pebble logs end early. After all, the process terminated, but the VM stayed up. Did pebble really not log anything about these at least two snapshot ingestions?

**Comment by tbg** [comment]
Either that or there is a lifecycle bug in CRDB, where we can remove a replica, block on pebble, add a new replica, then unblock the previous deletion and clobber it that way. It seems unlikely since there are a number of invariants that should have prevented it, but it's also code that we have worked on. See [thread](https://cockroachlabs.slack.com/archives/C054HQFC509/p1681990910249619?thread_ts=1681990491.897659&cid=C054HQFC509)

**Comment by jbowens** [comment]
Dumping the manifest ([MANIFEST-737943.zip](https://github.com/cockroachdb/cockroach/files/11286003/MANIFEST-737943.zip)):

```
5184171/12810
  log-num:       2105024
  next-file-num: 2105040
  last-seq-num:  11097369953
  added:         L0 2105038:82602<#11097368293-#11097369852>[/Local/RangeID/3/r/RangeAppliedState/0,0#11097369668,SET-/Table/42/13/2023-04-19T23:00:00Z/"tpce"/-0.5022153279999999/5/"f!70\x8bH\x00;"/"\xf2\xba\xd6&q\x11\x94\xc1"/"[[\xea\x7fd^\xc5\xd3"/4/0/1681948489.267298809,0#11097369319,SET] (2023-04-19T23:54:49Z)
5184320/12811
  log-num:       2105025
  next-file-num: 2105041
  last-seq-num:  11097369953
  added:         L0 2105018:1444<#11097369853-#11097369853>[/Local/RangeID/174089/r""/0,0#11097369853,RANGEKEYDEL-/Local/RangeID/174089/s""/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105023:1179<#11097369854-#11097369854>[/Local/RangeID/174089/u""/0,0#11097369854,RANGEDEL-/Local/RangeID/174089/v""/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105019:1833<#11097369855-#11097369855>[/Local/Range/Table/738/1/2183755417317264592/0,0#11097369855,RANGEKEYDEL-/Local/Range/Table/738/1/2312753627622925792/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105020:1248<#11097369856-#11097369856>[/Local/Lock/Intent/Local/Range/Table/738/1/2183755417317264592/0,0#11097369856,RANGEKEYDEL-/Local/Lock/Intent/Local/Range/Table/738/1/2312753627622925792/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105021:1224<#11097369857-#11097369857>[/Local/Lock/Intent/Table/738/1/2183755417317264592/0,0#11097369857,RANGEKEYDEL-/Local/Lock/Intent/Table/738/1/2312753627622925792/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L6 2105022:1200<#11097369858-#11097369858>[/Table/738/1/2183755417317264592/0,0#11097369858,RANGEKEYDEL-/Table/738/1/2312753627622925792/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
5185066/12812
  next-file-num: 2105042
  last-seq-num:  11097370020
  added:         L0 2105032:1445<#11097369961-#11097369961>[/Local/RangeID/174164/r""/0,0#11097369961,RANGEKEYDEL-/Local/RangeID/174164/s""/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105037:1179<#11097369962-#11097369962>[/Local/RangeID/174164/u""/0,0#11097369962,RANGEDEL-/Local/RangeID/174164/v""/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105033:2948<#11097369963-#11097369963>[/Local/Range/Table/738/1/5390282359200842992/0,0#11097369963,RANGEKEYDEL-/Local/Range/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105034:1248<#11097369964-#11097369964>[/Local/Lock/Intent/Local/Range/Table/738/1/5390282359200842992/0,0#11097369964,RANGEKEYDEL-/Local/Lock/Intent/Local/Range/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105035:1224<#11097369965-#11097369965>[/Local/Lock/Intent/Table/738/1/5390282359200842992/0,0#11097369965,RANGEKEYDEL-/Local/Lock/Intent/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L5 2105036:1200<#11097369966-#11097369966>[/Table/738/1/5390282359200842992/0,0#11097369966,RANGEKEYDEL-/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
```
There doesn't appear to have been a sequence number inversion here. The flush of the memtable to 2105038 is applied first, then the ingestion that ingests 2105037.

**Comment by tbg** [comment]
Assigning to storage.

Btw the keys involved in the assertion are here in env vars for convenience

```
grinaker-231-0002:/mnt/data1/cockroach-lease-state-fatal$ cat env.sh
# From assertion failure
BAD_RANGE=174164
RANGE_APPLIED_STATE_KEY=hex:0169f802a854727261736b00
RANGE_LEASE_KEY=hex:0169f802a85472726c6c2d00
TRUNCATED_STATE_KEY=hex:0169f802a854757266747400
HARD_STATE_KEY=hex:0169f802a854757266746800
REPLICA_ID_KEY=hex:0169f802a854757266747200
```

**Comment by jbowens** [comment]
An L0->L0 compaction compacted 2105038 and others into 2105042.

```
5186696/12816
  next-file-num: 2105044
  last-seq-num:  11097372171
  deleted:       L0 2104953
  deleted:       L0 2104954
  deleted:       L0 2104955
  deleted:       L0 2104956
  deleted:       L0 2104958
  deleted:       L0 2104960
  deleted:       L0 2104961
  deleted:       L0 2104962
  deleted:       L0 2104963
  deleted:       L0 2104965
  deleted:       L0 2104971
  deleted:       L0 2104973
  deleted:       L0 2104974
  deleted:       L0 2104975
  deleted:       L0 2104976
  deleted:       L0 2104978
  deleted:       L0 2104980
  deleted:       L0 2104981
  deleted:       L0 2104982
  deleted:       L0 2104983
  deleted:       L0 2104985
  deleted:       L0 2104987
  deleted:       L0 2104990
  deleted:       L0 2104991
  deleted:       L0 2104992
  deleted:       L0 2104993
  deleted:       L0 2104994
  deleted:       L0 2104996
  deleted:       L0 2104999
  deleted:       L0 2105000
  deleted:       L0 2105001
  deleted:       L0 2105002
  deleted:       L0 2105003
  deleted:       L0 2105005
  deleted:       L0 2105008
  deleted:       L0 2105009
  deleted:       L0 2105010
  deleted:       L0 2105011
  deleted:       L0 2105013
  deleted:       L0 2105014
  deleted:       L0 2105017
  deleted:       L0 2105018
  deleted:       L0 2105019
  deleted:       L0 2105020
  deleted:       L0 2105021
  deleted:       L0 2105023
  deleted:       L0 2105034
  deleted:       L0 2105038
  deleted:       L0 2105041
  added:         L0 2105042:364581<#11097361323-#11097369964>[/Local/RangeID/3/r/RangeAppliedState/0,0#11097369668,SET-/Table/42/13/2023-04-19T23:00:00Z/"tpce"/-0.32713391999999997/4/"\x11O\x02ĸ\x9an\x85"/"\x85\xd4\xe3\xd2B\xc3\xfa\x7f"/"\n\xe7\xa6419`x"/4/0/1681948489.200472888,0#11097367606,SET] (2023-04-19T23:54:49Z)
```

**Comment by jbowens** [comment]
And in the final state, 2105042 is in a higher sublevel than 2105037. 🙅 

```
--- L0.4 ---
  2105042:364581<#11097361323-#11097369964>[/Local/RangeID/3/r/RangeAppliedState/0,0#11097369668,SET-/Table/42/13/2023-04-19T23:00:00Z/"tpce"/-0.32713391999999997/4/"\x11O\x02ĸ\x9an\x85"/"\x85\xd4\xe3\xd2B\xc3\xfa\x7f"/"\n\xe7\xa6419`x"/4/0/1681948489.200472888,0#11097367606,SET]
--- L0.3 ---
  2105026:1445<#11097369948-#11097369948>[/Local/RangeID/174132/r""/0,0#11097369948,RANGEKEYDEL-/Local/RangeID/174132/s""/0,0#inf,RANGEDEL]
  2105031:1179<#11097369949-#11097369949>[/Local/RangeID/174132/u""/0,0#11097369949,RANGEDEL-/Local/RangeID/174132/v""/0,0#inf,RANGEDEL]
  2105032:1445<#11097369961-#11097369961>[/Local/RangeID/174164/r""/0,0#11097369961,RANGEKEYDEL-/Local/RangeID/174164/s""/0,0#inf,RANGEDEL]
  2105037:1179<#11097369962-#11097369962>[/Local/RangeID/174164/u""/0,0#11097369962,RANGEDEL-/Local/RangeID/174164/v""/0,0#inf,RANGEDEL]
  2105033:2948<#11097369963-#11097369963>[/Local/Range/Table/738/1/5390282359200842992/0,0#11097369963,RANGEKEYDEL-/Local/Range/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL]
  2105027:1828<#11097369950-#11097369950>[/Local/Range/Table/738/1/8725807511390082592/0,0#11097369950,RANGEKEYDEL-/Local/Range/Table/738/1/8799520774421888992/0,0#inf,RANGEDEL]
  2105028:1248<#11097369951-#11097369951>[/Local/Lock/Intent/Local/Range/Table/738/1/8725807511390082592/0,0#11097369951,RANGEKEYDEL-/Local/Lock/Intent/Local/Range/Table/738/1/8799520774421888992/0,0#inf,RANGEDEL]
  2105029:1224<#11097369952-#11097369952>[/Local/Lock/Intent/Table/738/1/8725807511390082592/0,0#11097369952,RANGEKEYDEL-/Local/Lock/Intent/Table/738/1/8799520774421888992/0,0#inf,RANGEDEL]
```

**Comment by itsbilal** [comment]
@jbowens where did you get the manifest? the zip you linked doesn't have `2105042`.

**Comment by jbowens** [comment]
@itsbilal — I forget, there's some nondeterminism in the sublevels right?

I'm not getting the same sublevels as `cockroach debug pebble manifest dump`.  eg, `L0.19` through manifest dump:

```
--- L0.19 ---
  2105026:1445<#11097369948-#11097369948>[/Local/RangeID/174132/r""/0,0#11097369948,RANGEKEYDEL-/Local/RangeID/174132/s""/0,0#inf,RANGEDEL]
  2105031:1179<#11097369949-#11097369949>[/Local/RangeID/174132/u""/0,0#11097369949,RANGEDEL-/Local/RangeID/174132/v""/0,0#inf,RANGEDEL]
  2105032:1445<#11097369961-#11097369961>[/Local/RangeID/174164/r""/0,0#11097369961,RANGEKEYDEL-/Local/RangeID/174164/s""/0,0#inf,RANGEDEL]
  2105037:1179<#11097369962-#11097369962>[/Local/RangeID/174164/u""/0,0#11097369962,RANGEDEL-/Local/RangeID/174164/v""/0,0#inf,RANGEDEL]
  2105033:2948<#11097369963-#11097369963>[/Local/Range/Table/738/1/5390282359200842992/0,0#11097369963,RANGEKEYDEL-/Local/Range/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL]
  2105027:1828<#11097369950-#11097369950>[/Local/Range/Table/738/1/8725807511390082592/0,0#11097369950,RANGEKEYDEL-/Local/Range/Table/738/1/8799520774421888992/0,0#inf,RANGEDEL]
  2105034:1248<#11097369964-#11097369964>[/Local/Lock/Intent/Local/Range/Table/738/1/5390282359200842992/0,0#11097369964,RANGEKEYDEL-/Local/Lock/Intent/Local/Range/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL]
  2105028:1248<#11097369951-#11097369951>[/Local/Lock/Intent/Local/Range/Table/738/1/8725807511390082592/0,0#11097369951,RANGEKEYDEL-/Local/Lock/Intent/Local/Range/Table/738/1/8799520774421888992/0,0#inf,RANGEDEL]
  2105035:1224<#11097369965-#11097369965>[/Local/Lock/Intent/Table/738/1/5390282359200842992/0,0#11097369965,RANGEKEYDEL-/Local/Lock/Intent/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL]
  2105029:1224<#11097369952-#11097369952>[/Local/Lock/Intent/Table/738/1/8725807511390082592/0,0#11097369952,RANGEKEYDEL-/Local/Lock/Intent/Table/738/1/8799520774421888992/0,0#inf,RANGEDEL]
```

whereas my unit test gets:
```
       0.19: file count: 5, bytes: 6924, width (mean, max): 1.0, 1, interval range: [33, 103]
        	2105026:[\x01i\xf8\x02\xa84r\x00#11097369948,19-\x01i\xf8\x02\xa84s\x00#72057594037927935,15]
        	2105031:[\x01i\xf8\x02\xa84u\x00#11097369949,15-\x01i\xf8\x02\xa84v\x00#72057594037927935,15]
        	2105027:[\x01k\x12\xf7\x02\xe2\x89\xfdy\x18K\xb3F\xd7V \x00\x01\x00#11097369950,19-\x01k\x12\xf7\x02\xe2\x89\xfdz\x1e-\x85\xc1\x16?\xe0\x00\x01\x00#72057594037927935,15]
        	2105028:[\x01zk\x12\x01k\x12\xf7\x02\xe2\x89\xfdy\x18K\xb3F\xd7V \x00\xff\x01\x00\x01\x00#11097369951,19-\x01zk\x12\x01k\x12\xf7\x02\xe2\x89\xfdz\x1e-\x85\xc1\x16?\xe0\x00\xff\x01\x00\x01\x00#72057594037927935,15]
        	2105029:[\x01zk\x12\xf7\x02\xe2\x89\xfdy\x18K\xb3F\xd7V \x00\x01\x00#11097369952,19-\x01zk\x12\xf7\x02\xe2\x89\xfdz\x1e-\x85\xc1\x16?\xe0\x00\x01\x00#72057594037927935,15]
```

**Comment by sumeerbhola** [comment]
>  I forget, there's some nondeterminism in the sublevels right?

shouldn't be any non-determinism (though I may be forgetting something)

**Comment by itsbilal** [comment]
@jbowens Maybe the incremental-generation of sublevels added some non-determinism. Is your unit test always generating sublevels from scratch? I think the manifest tool does the incremental generation whenever it can (i.e. using bve.Apply).

**Comment by jbowens** [comment]
So I think we've established that the L0->L0 compaction elevated 2105038's point tombstone 
```
    /Local/RangeID/174164/u/RaftTruncatedState/0,0#11097368379,DEL ""
 ```
 into a higher sublevel when it created 2105042.sst.
 
 But it also looks like it elevated 2104958's rangedel:
 ```
     /Local/RangeID/174164/u""/0,0-/Local/RangeID/174164/v""/0,0#11097361324,RANGEDEL
```

In Cockroach, I think the RANGEDEL is what caused the key to not be found (although the `DEL` is a problem too). I think in practice since the merging iterator merges across all levels (and we don't use `db.Get`) we're immune to problems from seqnum inversion of point tombstones unless a compaction drops something it shouldn't as a result of the inversion.

**Comment by jbowens** [comment]
> @jbowens Maybe the incremental-generation of sublevels added some non-determinism. Is your unit test always generating sublevels from scratch? I think the manifest tool does the incremental generation whenever it can (i.e. using bve.Apply).


Thanks, yeah, the tool `Accumulate`s into a single bve which it applies. My unit test was `Accumulate` and `Apply`-ing for every version edit. I think my unit test's behavior is more inline with what happens in production.

TODO: create an issue to update incremental sublevels calculation to be equivalent to a single batch recalculation.

<details>
just in case, if you use the `debug_test.go` unit test I linked up above and want the tool-equivalent sublevels construction, swap out `readCockroachManifest` for this

```
func readCockroachManifest(filename string) (*Version, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	rr := record.NewReader(f, 0 /* logNum */)
	var v *Version
	var bve BulkVersionEdit
	bve.AddedByFileNum = make(map[base.FileNum]*FileMetadata)
	for {
		r, err := rr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}
		var ve VersionEdit
		if err = ve.Decode(r); err != nil {
			return nil, err
		}
		if err := bve.Accumulate(&ve); err != nil {
			return nil, err
		}
		//if v, _, err = bve.Apply(v, EngineComparer.Compare, EngineComparer.FormatKey, 10<<20, 32000); err != nil {
		//return nil, err
		//}
	}
	v, _, err = bve.Apply(nil /* version */, EngineKeyCompare, EngineComparer.FormatKey, 0, 32000)
	return v, err
}
```
</details>

**Comment by jbowens** [comment]
Am I doing something silly in this unit test? https://github.com/cockroachdb/pebble/compare/crl-release-23.1...jbowens:pebble:23.1-debug

I was trying to induce this bug, but instead got
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x20 pc=0x102cf3af4]

goroutine 371 [running]:
github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).init(0x14002519180, {0x1031dc408?, 0x1400003c090}, 0x140005f2000, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}, ...)
	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/reader.go:416 +0x1e4
github.com/cockroachdb/pebble/sstable.(*Reader).NewIterWithBlockPropertyFiltersAndContext(0x140005f2000, {0x1031dc408, 0x1400003c090}, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}, 0x200000003?, ...)
	/Users/jackson/go/src/github.com/cockroachdb/pebble/sstable/reader.go:2657 +0x160
github.com/cockroachdb/pebble.(*tableCacheShard).newIters(0x140002300b0, {0x1031dc408, 0x1400003c090}, 0x140004bc2c0, 0x14000b8cc40, {0x0?, 0x1400087b290?, {0x0?, 0x0?}}, 0x14000266bd8)
	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:432 +0x3ec
github.com/cockroachdb/pebble.(*tableCacheContainer).newIters(0x14001161c98?, {0x1031dc408?, 0x1400003c090?}, 0x102db4b8c?, 0x0?, {0x0?, 0x1400087b290?, {0x0?, 0x0?}})
	/Users/jackson/go/src/github.com/cockroachdb/pebble/table_cache.go:139 +0x58
github.com/cockroachdb/pebble.(*levelIter).loadFile(0x14000b8cbe0, 0x140004bc2c0, 0x1)
	/Users/jackson/go/src/github.com/cockroachdb/pebble/level_iter.go:669 +0x238
github.com/cockroachdb/pebble.(*levelIter).Next(0x14000b8cbe0)
	/Users/jackson/go/src/github.com/cockroachdb/pebble/level_iter.go:882 +0xb8
github.com/cockroachdb/pebble.(*mergingIter).nextEntry(0x1400087b4c0, 0x14000b87480, {0x0?, 0x0?, 0x0?})
	/Users/jackson/go/src/github.com/cockroachdb/pebble/merging_iter.go:616 +0xe4
github.com/cockroachdb/pebble.(*mergingIter).findNextEntry(0x1400087b4c0)
	/Users/jackson/go/src/github.com/cockroachdb/pebble/merging_iter.go:800 +0xdc
github.com/cockroachdb/pebble.(*mergingIter).SeekGE(0x1400087b4c0, {0x140040d2564?, 0x8?, 0x140040d2564?}, 0x2?)
	/Users/jackson/go/src/github.com/cockroachdb/pebble/merging_iter.go:1109 +0xa4
github.com/cockroachdb/pebble.(*Iterator).SeekGEWithLimit(0x1400087af00, {0x140040d2564, 0x2, 0x2}, {0x0, 0x0, 0x0})
	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:1257 +0x58c
github.com/cockroachdb/pebble.(*Iterator).SeekGE(...)
	/Users/jackson/go/src/github.com/cockroachdb/pebble/iterator.go:1141
github.com/cockroachdb/pebble.TestIngestOverlapStress.func8()
	/Users/jackson/go/src/github.com/cockroachdb/pebble/ingest_test.go:2127 +0x110
created by github.com/cockroachdb/pebble.TestIngestOverlapStress
	/Users/jackson/go/src/github.com/cockroachdb/pebble/ingest_test.go:2121 +0xa20
exit status 2
FAIL	github.com/cockroachdb/pebble	0.562s
```

**Comment by bananabrick** [comment]
The test @jbowens posted panic even with flushable ingestions disabled. I'm debugging the test with flushable ingestions enabled.

The test is panicking because `r.readable` is nil: https://github.com/cockroachdb/pebble/blob/crl-release-23.1/sstable/reader.go#L416.

This is odd, because `r.readable` is set on every call to `NewReader`. I'm looking into the table cache to see why we would construct a Reader with a `nil` readable.

**Comment by RaduBerinde** [comment]
@bananabrick it's one of those red herrings, the panic happens earlier:

```
	/usr/local/go1.19.4/src/runtime/debug/stack.go:16 +0x19
github.com/cockroachdb/pebble/sstable.(*Reader).Close(0xc00183e600)
	/go/src/github.com/cockroachdb/pebble/sstable/reader.go:2608 +0x6a
github.com/cockroachdb/pebble.(*tableCacheValue).release(0xc00025faa0, 0xc0000c14a0)
	/go/src/github.com/cockroachdb/pebble/table_cache.go:964 +0x3f
github.com/cockroachdb/pebble.(*tableCacheShard).evict(0xc0000c14a0, 0xa0, 0xc00021e7e8, 0x1)
	/go/src/github.com/cockroachdb/pebble/table_cache.go:835 +0x265
github.com/cockroachdb/pebble.(*tableCacheShard).removeDB(0xc0000c14a0, 0xc00021e7e8)
	/go/src/github.com/cockroachdb/pebble/table_cache.go:867 +0x16e
github.com/cockroachdb/pebble.(*tableCacheContainer).close(0xc00021e7e0)
	/go/src/github.com/cockroachdb/pebble/table_cache.go:127 +0xed
github.com/cockroachdb/pebble.(*DB).Close(0xc00021b400)
	/go/src/github.com/cockroachdb/pebble/db.go:1451 +0x454
panic({0xbdbb00, 0xc00045cb80})
	/usr/local/go1.19.4/src/runtime/panic.go:884 +0x212
github.com/cockroachdb/pebble.randSpanInRange(0xc0005b4960?, 0x0?, 0x2bd)
	/go/src/github.com/cockroachdb/pebble/ingest_test.go:2227 +0x168
github.com/cockroachdb/pebble.TestIngestOverlapStress.func6(0xc0031c1bb8)
	/go/src/github.com/cockroachdb/pebble/ingest_test.go:2158 +0x274
github.com/cockroachdb/pebble.TestIngestOverlapStress(0xc0000fbba0)
	/go/src/github.com/cockroachdb/pebble/ingest_test.go:2196 +0xe49
```

**Comment by bananabrick** [comment]
@RaduBerinde Yea, made an edit. We're closing the reader while also concurrently trying to read from it. I don't understand why yet.

**Comment by RaduBerinde** [comment]
@bananabrick read the stack trace. It's because `randSpanInRange` in the test panics, and that leads to the deferred `db.Close` while the goroutines are still running.

**Comment by bananabrick** [comment]
Ahh, I see. Thanks for the help. I was hoping it was a problem in the test.

**Comment by sumeerbhola** [comment]
I called `intraL0CompactionUsingSeed` using `0.18: 2105041 [     2,   109]` with all the intervals in that range. All intra-L0 compactions include both 2105037 and 2105038. So we don't yet have a reproduction.

**Comment by sumeerbhola** [comment]
btw, `AddL0Files` (the incremental L0SubLevels initialization path) is being called in debug_test.go for the flush and the ingest, and the last edit before the one that picked the compaction (offset: 5186672) only deleted an L6 file, so it just reused the existing L0Sublevels.
```
    debug_test.go:68: offset: 5184171, called AddL0Files true
    debug_test.go:68: offset: 5184320, called AddL0Files true
    debug_test.go:68: offset: 5185066, called AddL0Files true
    debug_test.go:68: offset: 5185807, called AddL0Files true
    debug_test.go:68: offset: 5185926, called AddL0Files true
    debug_test.go:68: offset: 5186672, called AddL0Files false
```

**Comment by sumeerbhola** [comment]
I think we have the bug https://github.com/cockroachdb/pebble/blob/f87a15b23a26484db1b407de57db064ada26ca07/internal/manifest/l0_sublevels.go#L1568-L1576

File 2105037 is getting skipped there.

Sequence number ranges for files in order of sub-levels from higher to lower
```
2105041:<#11097369859-#11097369947> (the highest sub-level file in the compaction)
2105037:<#11097369962-#11097369962>
2105038:<#11097368293-#11097369852>
```
2105041 has a highest seqnum that is less than the highest seqnum of 2105037. So if we happen to have an `earliestUnflushedSeqNum` between those two, 2105041 and 2105038 will get picked and 2105037 will not get picked. This  reproduces easily.

I need to think whether this was always flawed or we've done something recently (like flushable ingests) that has broken some assumptions.

**Comment by itsbilal** [comment]
@sumeerbhola but then why did the output sstable have a highest seqnum of `11097369964` in it? We should have excluded `2105034` from the compaction because of the same `earliestUnflushedSeqNum` guard but we didn't.

**Comment by sumeerbhola** [comment]
Good point. I think it is because we have some implicit invariants when going down the sub-levels due to which we don't check the `earliestUnflushedSeqNum` for all adds. Specifically, this one https://github.com/cockroachdb/pebble/blob/f87a15b23a26484db1b407de57db064ada26ca07/internal/manifest/l0_sublevels.go#L1695. So if one calls `intraL0CompactionUsingSeed` with 2105041 and interval 75 (the interval of 2105034), 2105034 gets picked. This is reproducible:
```
seed compaction 75: 2105034=true, 2105037=false, 2105038=true
```

**Comment by itsbilal** [comment]
That invariant would hold if `2105034` were above `2105041` in sublevels. Looking at the key ranges and seqnum ranges, it absolutely should be, and it probably is if we regenerate sublevels. Except it isn't because, and this is likely the bug, `2105041` joins the LSM *after* `2105034` and takes the incremental path which assumes it must have the highest seqnums and just slaps it on top of `2105034`:

```
  next-file-num: 2105042
  last-seq-num:  11097370020
  added:         L0 2105032:1445<#11097369961-#11097369961>[/Local/RangeID/174164/r""/0,0#11097369961,RANGEKEYDEL-/Local/RangeID/174164/s""/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105037:1179<#11097369962-#11097369962>[/Local/RangeID/174164/u""/0,0#11097369962,RANGEDEL-/Local/RangeID/174164/v""/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105033:2948<#11097369963-#11097369963>[/Local/Range/Table/738/1/5390282359200842992/0,0#11097369963,RANGEKEYDEL-/Local/Range/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL] (
2023-04-19T23:54:49Z)
  added:         L0 2105034:1248<#11097369964-#11097369964>[/Local/Lock/Intent/Local/Range/Table/738/1/5390282359200842992/0,0#11097369964,RANGEKEYDEL-/Local/Lock/Intent/Local/Range/Table/738/1/542
7138990716746192/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L0 2105035:1224<#11097369965-#11097369965>[/Local/Lock/Intent/Table/738/1/5390282359200842992/0,0#11097369965,RANGEKEYDEL-/Local/Lock/Intent/Table/738/1/5427138990716746192/0,0#inf
,RANGEDEL] (2023-04-19T23:54:49Z)
  added:         L5 2105036:1200<#11097369966-#11097369966>[/Table/738/1/5390282359200842992/0,0#11097369966,RANGEKEYDEL-/Table/738/1/5427138990716746192/0,0#inf,RANGEDEL] (2023-04-19T23:54:49Z)
5185807/12813
  log-num:       2105039
  next-file-num: 2105042
  last-seq-num:  11097370306
  added:         L0 2105041:41219<#11097369859-#11097369947>[/Local/RangeID/3/u/RaftLog/logIndex:2521760/0,0#11097369860,SET-/Table/13/1/2023-04-19T23:54:49.301998Z/858122123028070402/6/1/168194848
9.301998237,0#11097369888,SET] (2023-04-19T23:54:49Z)
5185926/12814
```

The bug is likely here: https://github.com/cockroachdb/pebble/blob/f87a15b23a26484db1b407de57db064ada26ca07/internal/manifest/l0_sublevels.go#L663 ; we should be returning `errInvalidL0SublevelsOpt` regardless of what our `subLevel` is. The incoming file is sliding underneath existing files so we can't take the incremental approach to building sublevels. And then the `earliestUnflushedSeqNum` logic holds.

**Comment by sumeerbhola** [comment]
2105037 and 2105041 don't have any conflicting keys, since the latter came from a memtable that did not have any keys conflicting with the intent AND even though the memtable could later have added conflicting keys (with seqnum higher than that of 2105037) those keys are not in 2105041 since its seqnums are lower. 2105041 must be the result of a flush that split into multiple files. Given our L0=>Lbase picking logic, inverting the sublevels of 2105037 and 2105041 will be harmless, so I suspect this bug only affects intra-L0 compaction.
